src/herder/HerderImpl.cpp

```diff
@@ -1017,6 +1021,65 @@ HerderImpl::getCurrentlyTrackedQuorum() const
     return mPendingEnvelopes.getCurrentlyTrackedQuorum();
 }
 
+static Hash
+getQmapHash(QuorumTracker::QuorumMap const& qmap)
+{
+    std::unique_ptr<SHA256> hasher = SHA256::create();
+    for (auto const& pair : qmap)
+    {
+        hasher->add(xdr::xdr_to_opaque(pair.first));
+        if (pair.second)
+        {
+            hasher->add(xdr::xdr_to_opaque(*pair.second));
+        }
+        else
+        {
+            hasher->add("\0");
+        }
+    }
+    return hasher->finish();
+}
+
+void
+HerderImpl::checkAndMaybeReanalyzeQuorumMap()
+{
+    if (!mLastQuorumMapIntersectionState.mRecalculating)
+    {
+        QuorumTracker::QuorumMap const& qmap = getCurrentlyTrackedQuorum();
+        Hash curr = getQmapHash(qmap);
+        if (mLastQuorumMapIntersectionState.mLastCheckQuorumMapHash != curr)
+        {
+            CLOG(INFO, "Herder")
+                << "Transitive closure of quorum has changed, re-analyzing.";
+            mLastQuorumMapIntersectionState.mRecalculating = true;
+            auto& cfg = mApp.getConfig();
+            auto ledger = getCurrentLedgerSeq();
+            auto nNodes = qmap.size();
+            auto qic = QuorumIntersectionChecker::create(qmap, cfg);
+            auto& hState = mLastQuorumMapIntersectionState;
+            auto& app = mApp;
+            auto worker = [curr, ledger, nNodes, qic, &app, &hState] {
+                bool ok = qic->networkEnjoysQuorumIntersection();
+                auto split = qic->getPotentialSplit();
+                app.postOnMainThread(
```

Comment:
Is this safe on shutdown? In `~ApplicationImpl`, main IO service is stopped before the worker threads complete their tasks, so as this can run for a while, what happens when application shuts down before this line is executed?

Reply:
As far as I can tell from the asio docs, posting to a stopped io_service will return immediately without running the callback. Which isn't perfect -- the calculation will essentially remain half-completed and therefore unable to run again -- but it's only going to live another microsecond until the Application is torn down, so I don't think it's a big problem.